pkg/resourceinterpreter/configurableinterpreter/configurable.go

```diff
@@ -2,156 +2,327 @@ package configurableinterpreter
+func (c *ConfigurableInterpreter) Retain(desired *unstructured.Unstructured, observed *unstructured.Unstructured) (retained *unstructured.Unstructured, enabled bool, err error) {
+	eng, enabled := c.getInterpreter(desired.GroupVersionKind(), configv1alpha1.InterpreterOperationRetain)
+	if !enabled {
+		return
 	}
 
-	luaScript := customAccessor.GetRetentionLuaScript()
-	klog.Infof("lua script %s", luaScript)
-
-	return nil, err
+	rets, err := eng.Invoke(desired, observed)
+	if err != nil {
+		return
+	}
+	retained, err = toUnstructured(rets[0])
+	return
 }
 
 // AggregateStatus returns the objects that based on the 'object' but with status aggregated.
-func (c *ConfigurableInterpreter) AggregateStatus(object *unstructured.Unstructured, aggregatedStatusItems []workv1alpha2.AggregatedStatusItem) (*unstructured.Unstructured, error) {
-	klog.Infof("ConfigurableInterpreter Execute AggregateStatus")
-	customAccessor := c.configManager.LuaScriptAccessors()[object.GroupVersionKind()]
-	if len(customAccessor.GetStatusAggregationLuaScript()) == 0 {
-		return nil, fmt.Errorf("customized interpreter AggregateStatus Retain for %q not found", object.GroupVersionKind())
+func (c *ConfigurableInterpreter) AggregateStatus(object *unstructured.Unstructured, aggregatedStatusItems []workv1alpha2.AggregatedStatusItem) (obj *unstructured.Unstructured, enabled bool, err error) {
+	eng, enabled := c.getInterpreter(object.GroupVersionKind(), configv1alpha1.InterpreterOperationAggregateStatus)
+	if !enabled {
+		return
 	}
 
-	luaScript := customAccessor.GetStatusAggregationLuaScript()
-	klog.Infof("lua script %s", luaScript)
-
-	return nil, nil
+	rets, err := eng.Invoke(object, aggregatedStatusItems)
+	if err != nil {
+		return
+	}
+	obj, err = toUnstructured(rets[0])
+	return
 }
 
 // GetDependencies returns the dependent resources of the given object.
-func (c *ConfigurableInterpreter) GetDependencies(object *unstructured.Unstructured) (dependencies []configv1alpha1.DependentObjectReference, err error) {
-	klog.Infof("ConfigurableInterpreter Execute GetDependencies")
-
-	customAccessor := c.configManager.LuaScriptAccessors()[object.GroupVersionKind()]
-	if len(customAccessor.GetDependencyInterpretationLuaScript()) == 0 {
-		return nil, fmt.Errorf("customized interpreter GetDependencies Retain for %q not found", object.GroupVersionKind())
+func (c *ConfigurableInterpreter) GetDependencies(object *unstructured.Unstructured) (dependencies []configv1alpha1.DependentObjectReference, enabled bool, err error) {
+	eng, enabled := c.getInterpreter(object.GroupVersionKind(), configv1alpha1.InterpreterOperationInterpretDependency)
+	if !enabled {
+		return
 	}
 
-	luaScript := customAccessor.GetDependencyInterpretationLuaScript()
-	klog.Infof("lua script %s", luaScript)
+	rets, err := eng.Invoke(object)
+	if err != nil {
+		return
+	}
 
-	return nil, err
+	if r := rets[0]; !r.IsNil() {
+		err = r.Into(&dependencies)
+	}
```

Comment:
It seems not straightforward enough, since each operation may require a different number of inputs and returns.
I'm afraid it's going to be hard to maintain.

Reply:
The number of returns are declared here (the third arg in `Load` function)

https://github.com/karmada-io/karmada/blob/1c840eac2e52c36df20fbfc0893fd753c10c5491/pkg/resourceinterpreter/configurableinterpreter/configurable.go#L269

Reply:
The number of returns are declared here (the third arg in `Load` function)

https://github.com/karmada-io/karmada/blob/1c840eac2e52c36df20fbfc0893fd753c10c5491/pkg/resourceinterpreter/configurableinterpreter/configurable.go#L269

Reply:
In this place, the inputs and returns are certain. We can use it directly.